server/lib/chat/chat-models.ts

```diff
@@ -578,3 +579,94 @@ export async function findChannelByName(
     done()
   }
 }
+
+/** Returns the number of channels that match a particular condition. */
+async function getChannelsCount(
+  whereCondition?: SQLStatement,
+  withClient?: DbClient,
+): Promise<number> {
+  const { client, done } = await db(withClient)
+  try {
+    const query = sql`
+      SELECT COUNT(*) AS count
+      FROM channels
+    `
+
+    if (whereCondition) {
+      query.append(whereCondition)
+    }
+
+    const result = await client.query<{ count: string }>(query)
+
+    return Number(result.rows[0].count)
+  } finally {
+    done()
+  }
+}
+
+/**
+ * Gets a list of chat channels for a particular user, optionally filtered by a `searchStr`. If the
+ * user is an admin, this method will return a full list of channels. Otherwise, this method will
+ * only return non-private channels and channels that the user is not joined in.
+ */
+export async function getChannelsForUser(
+  userId: SbUserId,
+  isAdmin: boolean,
+  limit: number,
+  pageNumber: number,
+  searchStr?: string,
+  withClient?: DbClient,
+): Promise<{ channels: ChannelInfo[]; total: number }> {
+  const { client, done } = await db(withClient)
+  try {
+    const query = isAdmin
```

Comment:
I don't think this is really a great way to structure this functionality, it becomes rather unwieldy very quickly since the structure of these 2 queries is very different, but you need to add modifications based on things with the same name below. It also just becomes likely that you can break one and only test the other and not know/realize. I would split this into 2 functions: 1 for admin usage, 1 for "normal" usage. I don't think keeping these in the same function and reusing a few lines below is really buying you very much (and if you *really* can't deal with that duplication, you could move that logic into some function they both call into, but I do not think that's very worth it)